src/resolver.js

```diff
@@ -0,0 +1,32 @@
+let promises
+let resolving = false
+
+const performIteration = async (resolver, iterations, maxIterations = 0) => {
+  promises = []
+  const result = resolver()
+
+  if (iterations < maxIterations && promises.length > 0) {
+    await Promise.all(promises)
+    performIteration(resolver, iterations + 1, maxIterations)
+  }
+
+  return Promise.resolve(result)
+}
+
+export const registerRequestPromise = createRequestPromise => {
```

Comment:
Because the promise shouldn't be created unless we are actually resolving.still create a promise even if it shouldn't